<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro">

  <xacro:property name="eef" value="eef_2_finger"/>

  <xacro:arg name="serial_port_esp32" default="/dev/esp32_arm"/>
  <xacro:property name="serial_port" value="$(arg serial_port_esp32)"/>  
  <xacro:property name="baud_rate" value="115200"/>          <!-- default -->
  <xacro:property name="timeout" value="0.1"/>               <!-- default -->
  <!-- <xacro:property name="sim_mode" value="true"/> -->

  <xacro:unless value="$(arg sim_mode)">
    <ros2_control name="xarm_ros2_control" type="system">
      <hardware>
        <plugin>xarm_hardware/XArmHardware</plugin>
        <param name="example_param_hw_start_duration_sec">0</param>
        <param name="example_param_hw_stop_duration_sec">3.0</param>
        <param name="example_param_hw_slowdown">100</param>
        <param name="serial_port">${serial_port}</param>
        <param name="baud_rate">${baud_rate}</param>
        <param name="timeout">${timeout}</param>
      </hardware>

      <xacro:if value="${eef == 'eef_2_finger'}">
        <joint name="xarm_1_joint">
            <command_interface name="position">
            <param name="min">-1.832</param>
            <param name="max">0.75</param>
            </command_interface>
            <state_interface name="velocity"/>
            <state_interface name="position"/>
        </joint>

        <!-- mirror here -->

       <!-- <joint name="xarm_1_joint_mirror">

            <param name="mimic">xarm_1_joint</param>
            <param name="multiplier">-1</param>
            <state_interface name="position"/>
            <state_interface name="velocity"/>
        </joint> -->
      </xacro:if>

      <xacro:if value="${eef == 'eef_3_finger'}">
        <joint name="xarm_8_joint">
            <command_interface name="position">
            <param name="min">-0.36</param>
            <param name="max">0.44</param>
            </command_interface>
            <state_interface name="position"/>
            <state_interface name="velocity"/>
        </joint>

        <joint name="xarm_8_joint_mirror_1">
            <command_interface name="position">
            <param name="min">-0.36</param>
            <param name="max">0.44</param>
            </command_interface>
            <state_interface name="position"/>
            <state_interface name="velocity"/>
        </joint>

        <joint name="xarm_8_joint_mirror_2">
            <command_interface name="position">
            <param name="min">-0.36</param>
            <param name="max">0.44</param>
            </command_interface>
            <state_interface name="position"/>
            <state_interface name="velocity"/>
        </joint>
      </xacro:if>        

      <!-- <joint name="xarm_7_joint">
        <command_interface name="position">
          <param name="min">-1.57</param>
          <param name="max">1.57</param>
        </command_interface>
        <state_interface name="position"/>
      </joint> -->

      <joint name="xarm_2_joint">
        <command_interface name="position">
          <param name="min">-2.35</param>
          <param name="max">2.35</param>
        </command_interface>
        <state_interface name="position"/>
        <state_interface name="velocity"/>
      </joint>

      <joint name="xarm_3_joint">
        <command_interface name="position">
          <param name="min">-2.05</param>
          <param name="max">2.35</param>
        </command_interface>
        <state_interface name="position"/>
        <state_interface name="velocity"/>
      </joint>

      <joint name="xarm_4_joint">
        <command_interface name="position">
          <param name="min">-2.2</param>
          <param name="max">2.2</param>
        </command_interface>
        <state_interface name="position"/>
        <state_interface name="velocity"/>
      </joint>

      <joint name="xarm_5_joint">
        <command_interface name="position">
          <param name="min">-1.62</param>
          <param name="max">1.62</param>
        </command_interface>
        <state_interface name="position"/>
        <state_interface name="velocity"/>
      </joint>

      <joint name="xarm_6_joint">
        <command_interface name="position">
          <param name="min">-2.35</param>
          <param name="max">2.35</param>
        </command_interface>
        <state_interface name="position"/>
        <state_interface name="velocity"/>
      </joint>
    </ros2_control>
  </xacro:unless>

  <xacro:if value="$(arg sim_mode)">
    <ros2_control name="arm_system" type="system">
        <hardware>
            <plugin>gz_ros2_control/GazeboSimSystem</plugin>
        </hardware>
        <xacro:if value="${eef == 'eef_2_finger'}">
        <joint name="xarm_1_joint">
            <command_interface name="position">
            <param name="min">0.0035</param>
            <param name="max">0.028</param>
            </command_interface>
            <state_interface name="position">
                <param name="initial_value">0.013</param>
            </state_interface>
            <state_interface name="velocity"/>
        </joint>

        <joint name="xarm_1_joint_mirror">
            <!-- <param name="mimic">
                <joint>xarm_1_joint</joint>
                <multiplier>-1</multiplier>
            </param> -->
            <param name="mimic">xarm_1_joint</param>
            <param name="multiplier">-1</param>
            <state_interface name="position"/>
            <state_interface name="velocity"/>
        </joint>

        <!-- <joint name="xarm_1_joint_mirror">
            <command_interface name="position">
            <param name="min">0.003</param>
            <param name="max">0.028</param>
            </command_interface>
            <state_interface name="position">
                <param name="initial_value">0.003</param>
            </state_interface>
            <state_interface name="velocity"/>
        </joint> -->
      </xacro:if>

      <xacro:if value="${eef == 'eef_3_finger'}">
        <joint name="xarm_8_joint">
            <command_interface name="position">
            <param name="min">-0.36</param>
            <param name="max">0.44</param>
            </command_interface>
            <state_interface name="position"/>
            <state_interface name="velocity"/>
        </joint>

        <joint name="xarm_8_joint_mirror_1">
            <command_interface name="position">
            <param name="min">-0.36</param>
            <param name="max">0.44</param>
            </command_interface>
            <state_interface name="position"/>
            <state_interface name="velocity"/>
        </joint>

        <joint name="xarm_8_joint_mirror_2">
            <command_interface name="position">
            <param name="min">-0.36</param>
            <param name="max">0.44</param>
            </command_interface>
            <state_interface name="position"/>
            <state_interface name="velocity"/>
        </joint>
      </xacro:if>        

      <!-- <joint name="xarm_7_joint">
        <command_interface name="position">
          <param name="min">-1.57</param>
          <param name="max">1.57</param>
        </command_interface>
        <state_interface name="position"/>
      </joint> -->

      <joint name="xarm_2_joint">
        <command_interface name="position">
          <param name="min">-1.208</param>
          <param name="max">1.933</param>
        </command_interface>
        <state_interface name="position"/>
        <state_interface name="velocity"/>
      </joint>

      <joint name="xarm_3_joint">
        <command_interface name="position">
          <param name="min">-1.613</param>
          <param name="max">1.528</param>
        </command_interface>
        <state_interface name="position"/>
        <state_interface name="velocity"/>
      </joint>

      <joint name="xarm_4_joint">
        <command_interface name="position">
          <param name="min">-1.571</param>
          <param name="max">1.571</param>
        </command_interface>
        <state_interface name="position"/>
        <state_interface name="velocity"/>
      </joint>

      <joint name="xarm_5_joint">
        <command_interface name="position">
          <param name="min">-1.613</param>
          <param name="max">1.528</param>
        </command_interface>
        <state_interface name="position"/>
        <state_interface name="velocity"/>
      </joint>

      <joint name="xarm_6_joint">
        <command_interface name="position">
          <param name="min">-1.706</param>
          <param name="max">1.436</param>
        </command_interface>
        <state_interface name="position"/>
        <state_interface name="velocity"/>
      </joint>
    </ros2_control>
  </xacro:if>

  <!-- <gazebo>
      <plugin filename="gz_ros2_control-system" name="gz_ros2_control::GazeboSimROS2ControlPlugin"> -->
      <!-- <plugin name="gz_ros2_control::GazeboSimROS2ControlPlugin" filename="libgz_ros2_control-system.so"> -->
          <!-- <parameters>$(find xarm_description)/config/xarm_controllers.yaml</parameters>
          <parameters>$(find xarm_description)/config/gaz_ros2_ctl_use_sim.yaml</parameters>
      </plugin>
  </gazebo> -->

</robot>